aea/cli/packages.py

```diff
@@ -0,0 +1,376 @@
+
+
+class PackageManager:
+    """AEA package manager"""
+
+    path: Path
+    _packages: OrderedDictType[PackageId, str]
+
+    def __init__(
+        self, path: Path, packages: Optional[OrderedDictType[PackageId, str]] = None
+    ) -> None:
+        """Initialize object."""
+
+        self.path = path
+        self._packages_file = path / PACKAGES_FILE
+        self._packages = packages or OrderedDict()
+
+        self._logger = logging.getLogger(name="PackageManager")
+        self._logger.setLevel(logging.INFO)
+
+    @property
+    def packages(
+        self,
+    ) -> OrderedDictType[PackageId, str]:
+        """Returns mappings of package ids -> package hash"""
+
+        return self._packages
+
+    def sync(
+        self,
+        update_packages: bool = False,
+        update_hashes: bool = False,
+    ) -> "PackageManager":
+        """
+        Sync local packages to the remote registry.
+
+        :param update_packages: Update packages if the calculated hash for a
+                                package does not match the one in the packages.json.
+        :param update_hashes: Update hashes in the packages.json if the calculated
+                              hash for a package does not match the one in the
+                              packages.json.
+        :return: PackageManager object
+        """
+
+        self._logger.info(f"Performing sync @ {self.path}")
+        sync_needed = False
+
+        for package_id in self.packages:
+            package_path = self.package_path_from_package_id(package_id)
+
+            if package_path.exists():
+                package_hash = IPFSHashOnly.get(str(package_path))
+                expected_hash = self.packages[package_id]
+
+                if package_hash == expected_hash:
+                    continue
+
+                if update_hashes:
+                    self._logger.info(f"Updating hash for {package_id}")
+                    self._packages[package_id] = package_hash
+                    continue
+
+                if update_packages:
+                    sync_needed = True
+                    self._logger.info(
+                        f"Hash does not match for {package_id}, downloading package again..."
+                    )
+                    self.update_package(
+                        package_path, package_id.with_hash(expected_hash)
+                    )
+                    continue
+
+                raise PackageHashDoesNotMatch(
+                    f"Hashes for {package_id} does not match; Calculated hash: {package_hash}; Expected hash: {expected_hash}"
+                )
+
+            self._logger.info(f"{package_id} not found locally, downloading...")
+            package_id_with_hash = package_id.with_hash(self.packages[package_id])
+            self.add_package(package_id=package_id_with_hash)
+
+        if sync_needed:
+            self._logger.info("Sync complete")
+        else:
+            self._logger.info("No package was updated.")
```

Comment:
minor: if no package get updated but we download those that do not exist we also log that no package was updated